File: providers.py
#!/usr/bin/env python3
"""Read providers.json, and resolve a provider id to the set of files that belong to it.

Thirty lines rather than a class hierarchy, because there is exactly one thing to answer: given a
provider id, which manifest and which asset tree. Every Python tool in this repository — verify.py,
derive.py, sheet.py, normalise_ground.py, compare.py — takes `--provider` and asks that question
here, so none of them contains a path built by hand.

    python3 providers.py            # what is registered, and which sets exist on disk
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass
from pathlib import Path

HERE = Path(__file__).resolve().parent
REGISTRY = HERE / "providers.json"


@dataclass(frozen=True)
class Provider:
    id: str
    label: str
    vendor: str
    adapter: str
    #: Which prompt dialect this set was generated in; the rules are in dialects.json and the
    #: transform is dialects.py. `adapter` says how a set was POSTED, this says what was posted.
    #: Two providers can share a deployment, a route, a key and a model and still be two different
    #: experiments; the registry held exactly such a pair while the Qwen challenger was on trial,
    #: and both went when the owner withdrew that model. Parity is enforced WITHIN a dialect and
    #: RE-DERIVED across dialects; see verify.py's check_parity.
    dialect: str
    #: Repository-relative root. "." for the reference provider; see providers.json's header.
    root: Path
    #: "live" or "withdrawn". A withdrawn provider's deployment no longer exists; its entry is kept
    #: because the wire facts in it were measured. Nothing here counts providers.
    status: str
    shipped: bool
    implemented: bool
    concurrency: int
    billing: dict
    env: dict

    @property
    def assets(self) -> Path:
        return self.root / "assets"

    @property
    def manifest(self) -> Path:
        return self.root / "MANIFEST.json"

    @property
    def deployment(self) -> Path:
        """Where the operator records when this deployment existed. Only meaningful hourly."""
        return self.root / "DEPLOYMENT.json"

    @property
    def exists(self) -> bool:
        return self.manifest.exists()


def _document() -> dict:
    return json.loads(REGISTRY.read_text())


def key_of(entry: dict) -> str:
    """The identity of one asset, spelled the way THIS repository spells it.

    Driven by providers.json's `identity` block rather than hardcoded, because it is the one thing
    that really differs between the estate's three asset repositories — micro-brand keys on
    surface + kind, the two game sets key on a single dotted `asset` path. Everything else in this
    file is identical across all three, and stays that way because of this function.
    """
    fields = _document()["identity"]["key"]
    return "/".join(str(entry[f]) for f in fields[:-1]) + "@" + str(entry[fields[-1]])


def label_of(entry: dict) -> str:
    """The short human name for one asset, for a report line or a sheet caption."""
    return "/".join(str(entry[f]) for f in _document()["identity"]["label"])


def load() -> list[Provider]:
    document = _document()
    out = []
    for raw in document["providers"]:
        out.append(
            Provider(
                id=raw["id"],
                label=raw["label"],
                vendor=raw["vendor"],
                adapter=raw["adapter"],
                # No default. A provider with no declared dialect is a set nobody can say what was
                # sent to, and defaulting it to "literal" would silently claim a set is part of the
                # controlled comparison when the person who added it never said so.
                dialect=raw["dialect"],
                root=(HERE / raw["root"]).resolve(),
                status=raw.get("status", "live"),
                shipped=bool(raw["shipped"]),
                implemented=bool(raw["implemented"]),
                concurrency=int(raw["concurrency"]),
                billing=raw["billing"],
                env=raw["env"],
            )
        )
    return out


def reference() -> Provider:
    """The provider whose set every other set is judged against, and replays the prompts of."""
    wanted = _document()["reference"]
    return by_id(wanted)


def by_id(provider_id: str) -> Provider:
    for provider in load():
        if provider.id == provider_id:
            return provider
    known = ", ".join(p.id for p in load())
    raise SystemExit(f"unknown provider {provider_id!r}; providers.json registers: {known}")


def live() -> list[Provider]:
    """Every provider that can be run against today. Never a hardcoded pair."""
    return [p for p in load() if p.status == "live"]


def in_dialect(dialect: str) -> list[Provider]:
    """The providers sharing one prompt dialect — the group prompt parity is asserted over.

    Before there was one implicit group holding every provider. There are now N explicit ones, each
    holding the same property, and nothing here assumes a group has one member or that there is one
    group.
    """
    return [p for p in load() if p.dialect == dialect]


def present() -> list[Provider]:
    """Every registered provider that actually has a manifest on disk.

    This is what the tools default to. A run of verify.py or compare.py must not fail merely
    because a challenger has not been generated yet — that is the normal state, not a fault, and it
    is the state a withdrawn provider is permanently in.
    """
    return [p for p in load() if p.exists]


def add_argument(parser) -> None:
    """The one flag, spelled the same way by every tool here."""
    parser.add_argument(
        "--provider",
        action="append",
        default=None,
        metavar="ID",
        help="provider id from providers.json; repeatable. Default: every set present on disk.",
    )


def selected(args) -> list[Provider]:
    if not getattr(args, "provider", None):
        return present()
    return [by_id(value) for value in args.provider]


if __name__ == "__main__":
    for provider in load():
        state = "present" if provider.exists else "not generated"
        shipped = "shipped" if provider.shipped else "candidate"
        adapter = "implemented" if provider.implemented else "STUB"
        print(
            f"{provider.id:<26} {provider.dialect:<9} {provider.status:<10} {shipped:<9} "
            f"{adapter:<11} {state:<13} "
            f"{provider.root.relative_to(HERE) if provider.root != HERE else '.'}"
        )
    sys.exit(0)

File: providers.ts
/**
 * The TypeScript half of the provider registry. Reads the same `providers.json` `providers.py`
 * reads, so the two halves cannot disagree about where a set lives or what it costs.
 *
 * Nothing here knows how to CALL a provider — that is `backends.ts`. This file only answers "which
 * providers exist, which one is the reference, and which files belong to each", which is the
 * question every tool in the repository has to answer before it can do anything at all.
 */

import { readFileSync } from 'node:fs'
import { join, resolve } from 'node:path'

const HERE = import.meta.dirname

export type AdapterKind =
  | 'foundry-serverless'
  /** `/managed-deployments/<name>/v1/chat/completions`. Real, on another host; see backends.ts. */
  | 'foundry-managed-compute'
  /**
   * `/openai/v1/images/generations` on an Azure AI Foundry AIServices resource — the OpenAI images
   * envelope, `api-key` auth, `b64_json` in the response and no `url` field.
   *
   * This member existed once before and was deleted with the Qwen challenger, along with that
   * endpoint's transposed-`size` workaround. It is back because a second model is on trial again,
   * and it is a DIFFERENT implementation of the same shape: the envelope is the vendor's, the bugs
   * being compensated for are not. See `openaiImagesBackend` in backends.ts for the measured
   * contract, all of which was established by probing rather than read off a model card.
   */
  | 'openai-images'

/**
 * Whether a provider can be run against at all.
 *
 * `withdrawn` is not `implemented: false`. An unimplemented backend is one whose wire shape is not
 * known yet; a withdrawn one is a deployment that no longer exists. Cosmos 3 Super failed to come
 * up on A100_80GB and was deleted, and its entry is kept complete rather than removed — the estate
 * has a stated 3D/animation gap FLUX cannot fill (docs/ecosystem/19-new-products.md), so a third
 * model is a question of when. Nothing in this repository counts providers or assumes there are
 * two: the registry is the list, and `live()` is the subset that can be run.
 */
export type ProviderStatus = 'live' | 'withdrawn'

export interface ProviderBilling {
  /** The unit this model actually bills in. Never converted; see COMPARISON.md §6. */
  readonly unit: string
  readonly basis: string
  readonly source: string
  readonly sku: string | null
  readonly hourlyRate: number | null
}

export interface Provider {
  readonly id: string
  readonly label: string
  readonly vendor: string
  readonly adapter: AdapterKind
  /**
   * Which prompt dialect this set was generated in — `dialects.json`, read through `dialects.ts`.
   *
   * `adapter` says how a set was POSTED; this says what was posted. Two providers may share a
   * deployment, a route, a key and a model and still be two different experiments — identical in
   * every wire fact, different in this one field. The registry held exactly such a pair while the
   * Qwen challenger was on trial; both entries went when the owner withdrew that model, and the
   * machinery that made the pair expressible is kept because the next challenger will need it.
   *
   * Parity is enforced WITHIN a dialect and re-derived ACROSS dialects; see `dialects.ts`'s header
   * for why that is stronger than the plain equality it replaces, rather than weaker.
   */
  readonly dialect: string
  /** Absolute. `assets/` and `MANIFEST.json` hang off this. */
  readonly root: string
  readonly status: ProviderStatus
  readonly shipped: boolean
  /** False for a backend that is still a stub. `backendFor` throws rather than guessing a body. */
  readonly implemented: boolean
  readonly concurrency: number
  readonly env: Readonly<Record<string, string>>
  readonly billing: ProviderBilling
  /**
   * Managed Compute only: the deployment name that goes in the route. One host serves both
   * candidates, so the name is what distinguishes them.
   */
  readonly deployment?: string
  /**
   * False where the deployment name above is an assumption rather than a measured fact. Seven of
   * the eight FLUX model names probed on the reference resource answered 404, so an unverified
   * name is a real risk and not a formality. `probe.ts` is what turns this true.
   */
  readonly deploymentVerified?: boolean
  /** `/managed-deployments/{deployment}/v1/chat/completions`. */
  readonly route?: string
}

interface RegistryDocument {
  readonly reference: string
  readonly providers: ReadonlyArray<Omit<Provider, 'root'> & { readonly root: string }>
}

const document = JSON.parse(readFileSync(join(HERE, 'providers.json'), 'utf8')) as RegistryDocument

export const PROVIDERS: readonly Provider[] = document.providers.map((raw) => ({
  ...raw,
  root: resolve(HERE, raw.root),
}))

export function providerById(id: string): Provider {
  const found = PROVIDERS.find((p) => p.id === id)
  if (!found) {
    throw new Error(
      `unknown provider "${id}"; providers.json registers: ${PROVIDERS.map((p) => p.id).join(', ')}`,
    )
  }
  return found
}

/**
 * The provider every other set is judged against — and, more importantly, the one whose recorded
 * prompts a candidate run REPLAYS rather than recomputes. See generate.ts's prompt-parity header.
 */
export const REFERENCE: Provider = providerById(document.reference)

/** Every provider that can actually be run against today. Never a hardcoded pair. */
export const live = (): readonly Provider[] => PROVIDERS.filter((p) => p.status === 'live')

/**
 * The providers sharing one dialect — the group parity is asserted over.
 *
 * This is the seam that lets a second prompt dialect exist without weakening the guarantee the
 * comparison rests on. Before there was one implicit group containing every provider; there are now
 * N explicit ones, each holding the same property, and no code anywhere assumes a group has one
 * member or that there is one group.
 */
export const inDialect = (dialect: string): readonly Provider[] =>
  PROVIDERS.filter((p) => p.dialect === dialect)

/** Every dialect at least one registered provider is actually generated in. */
export const dialectsInUse = (): readonly string[] => [...new Set(PROVIDERS.map((p) => p.dialect))]

export class ProviderWithdrawnError extends Error {
  constructor(provider: Provider) {
    super(
      `${provider.id} is withdrawn: its deployment no longer exists, so there is nothing to run ` +
        'against. Its registry entry is kept because the wire facts in it were measured and are ' +
        'cheaper to re-read than to re-establish. Redeploy it and set status to "live".',
    )
    this.name = 'ProviderWithdrawnError'
  }
}

export const assetsDirOf = (provider: Provider): string => join(provider.root, 'assets')
export const manifestPathOf = (provider: Provider): string => join(provider.root, 'MANIFEST.json')
